src/core/init/initialize_media_source.ts

```diff
@@ -425,6 +427,20 @@ export default function InitializeOnMediaSource(
               case "protected-segment":
                 protectedSegments$.next(evt.value);
                 break;
+              case "adaptationChange":
+                if (evt.value.type === "audio" && evt.value.sourceBufferStatus === "initialized") {
+                  const { audioTrackSwitchingMode } = bufferOptions;
+                  if (
+                    audioTrackSwitchingMode === "flush" &&
+                    mediaElement.currentTime + 0.001 < mediaElement.duration
+                    ) {
```

Comment:
There is technically a risk of a race condition here, no?

Basically the term is "flush" here is because a seek will often (on most platforms) trigger on the low-level side a re-constructing of the whole buffer based on what we pushed.

This is sometimes necessary because when switching the audio track, we might have to replace the segments linked to the old audio track. The platform is not always completely synchronized with the `SourceBuffers`, as several layers of buffers exist until video and audio are decode. This could mean that while the JS-side SourceBuffers contain the new audio segments, the buffers used for decoding still contain the segments from the previous audio track.
When we seek, the lower-level code often run a logic which completely refresh their internal buffer's status: old segments are dropped and new segments are considered.

So why do I talk about a race condition here? Basically the idea is that we should seek after modifications have been done on the SourceBuffer (I would guess doing it after removing the previous segments is sufficient - though I'm not sure). Here, both the removal and the seek happens concurrently (the removal in the `PeriodStream` the seek here), without any synchronization between the two.
Also because of that, I'm not sure that  there is the effect we look for. We might want to test on ChromeCast to be sure.

Reply:
Oh, I perfectly see the potential race condition here! It makes sense now!

I just realize now that there is a lot of timing inside a video player and an adaptation change doesn't necessarily mean that the new segments are already downloaded and pushed inside the buffer. And so, we only need to do a seek once the new segments are added to the audio buffer.

Thanks for spotting this, I guess its a nooby mistake hahaha, I will reiterating on this!

Reply:
No problem, it's a kind of mistake we also do very often.

> an adaptation change doesn't necessarily mean that the new segments are already downloaded and pushed inside the buffer

Yes I think the key thing here is to not rely on implicit assumptions from events. For example an `adaptationChange` event is today documented as `Emitted when a new AdaptationStream is created...`. This says nothing about if old segments have been removed or if new ones have been pushed.
Consequently, this event should only be used to notify that a new Adaptation is now considered.

This becomes more important when such assumptions are true as of now.
If due to our code evolving they become false in the future, we wouldn't necessarily know we did something wrong if the corresponding event is still sent for what it is documented but we would create regressions in other parts of our code.